eryx/pdb.py

```diff
@@ -316,3 +341,308 @@ def _tile_form_factors(self):
+            - dimension 4: index atoms in neighbor ASU
+        """
+        hessian = np.zeros((self.n_asu, self.n_atoms_per_asu,
+                            self.n_cell, self.n_asu, self.n_atoms_per_asu),
+                           dtype='complex')
+        hessian_diagonal = np.zeros((self.n_asu, self.n_atoms_per_asu),
+                                    dtype='complex')
+
+        # off-diagonal
+        for i_asu in range(self.n_asu):
+            for i_cell in range(self.n_cell):
+                for j_asu in range(self.n_asu):
+                    for i_at in range(self.n_atoms_per_asu):
+                        iat_neighbors = self.asu_neighbors[i_asu][i_cell][j_asu][i_at]
+                        if len(iat_neighbors) > 0:
+                            hessian[i_asu, i_at, i_cell, j_asu, iat_neighbors] = -self.gamma[i_cell, i_asu, j_asu]
+                            hessian_diagonal[i_asu, i_at] -= self.gamma[i_cell, i_asu, j_asu] * len(iat_neighbors)
+
+        # diagonal (also correct for over-counted self term)
+        for i_asu in range(self.n_asu):
+            for i_at in range(self.n_atoms_per_asu):
+                hessian[i_asu, i_at, self.id_cell_ref, i_asu, i_at] = -hessian_diagonal[i_asu, i_at] - self.gamma[
+                    self.id_cell_ref, i_asu, i_asu]
+
+        return hessian
+
+    def compute_K(self, hessian, kvec=None):
+        """
+        Noting H(d) the block of the hessian matrix
+        corresponding the the d-th reference cell
+        whose origin is located at r_d, then:
+        K(kvec) = \sum_d H(d) exp(i kvec. r_d)
+
+        Parameters
+        ----------
+        hessian : numpy.ndarray, see compute_hessian()
+        kvec : numpy.ndarray, shape (3,)
+            phonon wavevector, default array([0.,0.,0.])
+
+        Returns
+        -------
+        Kmat : numpy.ndarray,
+               shape (n_asu, n_atoms_per_asu,
+                      n_asu, n_atoms_per_asu)
+               type 'complex'
+        """
+        if kvec is None:
+            kvec = np.zeros(3)
+        Kmat = hessian[:, :, self.id_cell_ref, :, :]
+
+        for j_cell in range(self.n_cell):
+            if j_cell == self.id_cell_ref:
+                continue
+            r_cell = self.crystal.get_unitcell_origin(self.crystal.id_to_hkl(j_cell))
+            phase = np.dot(kvec, r_cell)
+            eikr = np.cos(phase) + 1j * np.sin(phase)
+            for i_asu in range(self.n_asu):
+                for j_asu in range(self.n_asu):
+                    Kmat[i_asu, :, j_asu, :] += hessian[i_asu, :, j_cell, j_asu, :] * eikr
+        return Kmat
+
+    def compute_Kinv(self, hessian, kvec=None):
+        """
+        Compute the inverse of K(kvec)
+        (see compute_K() for the relationship between K and the hessian).
+
+        Parameters
+        ----------
+        hessian : numpy.ndarray, see compute_hessian()
+        kvec : numpy.ndarray, shape (3,)
+            phonon wavevector, default array([0.,0.,0.])
```

Comment:
For my understanding: a kvec of (0,0,0) implies no lattice dynamics? (And not that this is being evaluated at an integral Miller index?)

Reply:
Excellent question, let me see if I understand it enough to detail things a bit below:

Here, `kvec=(0,0,0)` is an arbitrary choice for a default vector. 

In the case where we consider interactions across asymmetric units, it does correspond to the wavevector of acoustic phonons and more particularly to translations of the crystal as a whole (longest wavelength), which indeed does not change interatomic distances and only contribute at integral Miller indices.

In the NIDM model, pairwise interactions across unit cells are zero in the first place. This is implemented by setting the corresponding entries in the hessian matrix $\mathbf{H}$ to zero. Then, the matrix 
```math
\mathbf{K}(\mathbf{k}) = \sum_{c} \mathbf{H}_{0,c} e^{i\mathbf{k}^T \mathbf{r}}
```
reduces to just the Hessian of the reference unit cell $\mathbf{H_{0,0}}$ and thus independent of $\mathbf{k}$. So in the NIDM model we could actually compute the generalized inverse of $\mathbf{K}$ at any wavevector $\mathbf{k}$ and get the same result: the covariance matrix - which will be zero for every pair of atoms that do not belong to the same unit cell.

Reply:
Got it. Thanks for that explanation!